lib/Transforms/LLZKPolyLoweringPass.cpp

```diff
@@ -0,0 +1,477 @@
+//===-- LLZKPolyLoweringPass.cpp ------------*- C++ -*-===//
+//
+// Part of the LLZK Project, under the Apache License v2.0.
+// See LICENSE.txt for license information.
+// Copyright 2025 Veridise Inc.
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+///
+/// \file
+/// This file implements the `-llzk-poly-lowering` pass.
+///
+//===----------------------------------------------------------------------===//
+
+#include "llzk/Dialect/Array/IR/Ops.h"
+#include "llzk/Dialect/Constrain/IR/Ops.h"
+#include "llzk/Dialect/Felt/IR/Ops.h"
+#include "llzk/Dialect/Function/IR/Ops.h"
+#include "llzk/Transforms/LLZKTransformationPasses.h"
+
+#include <mlir/IR/BuiltinOps.h>
+
+#include <llvm/ADT/DenseMap.h>
+#include <llvm/ADT/DenseMapInfo.h>
+#include <llvm/ADT/SmallVector.h>
+#include <llvm/Support/Debug.h>
+
+#include <deque>
+#include <memory>
+
+// Include the generated base pass class definitions.
+namespace llzk {
+#define GEN_PASS_DECL_POLYLOWERINGPASS
+#define GEN_PASS_DEF_POLYLOWERINGPASS
+#include "llzk/Transforms/LLZKTransformationPasses.h.inc"
+} // namespace llzk
+
+using namespace mlir;
+using namespace llzk;
+using namespace llzk::felt;
+using namespace llzk::function;
+using namespace llzk::component;
+using namespace llzk::constrain;
+
+#define DEBUG_TYPE "llzk-poly-lowering-pass"
+#define AUXILIARY_FIELD_PREFIX "poly_split_"
+
+namespace {
+
+struct AuxAssignment {
+  std::string auxFieldName;
+  Value computedValue;
+};
+
+class PolyLoweringPass : public llzk::impl::PolyLoweringPassBase<PolyLoweringPass> {
+private:
+  unsigned auxCounter = 0;
+
+  void collectStructDefs(ModuleOp modOp, SmallVectorImpl<StructDefOp> &structDefs) {
+    modOp.walk([&](StructDefOp structDef) { structDefs.push_back(structDef); });
```

Comment:
This lambda should return `WalkResult::skip()` so it doesn't recurse into the `StructDefOp` in search for nested `StructDefOp`s, which won't exist.